crates/warg/tests/package.rs

```diff
@@ -1,81 +1,105 @@
-use std::fs::{self, DirEntry};
-
-use hashbrown::HashMap;
+use anyhow::{anyhow, Context, Result};
 use pretty_assertions::assert_eq;
-use semver::Version;
 use serde::{Deserialize, Serialize};
-use serde_json;
+use std::{
+    fs::{self, DirEntry},
+    path::Path,
+};
 use warg_protocol::{
     hash,
-    package::{self, validate::ValidationState},
+    package::{self, validate::Validator},
     protobuf, signing, Envelope,
 };
 
 #[test]
-fn test_package_logs() {
-    let mut entries: Vec<DirEntry> = fs::read_dir("./tests/package-logs")
-        .unwrap()
-        .collect::<Result<Vec<DirEntry>, _>>()
-        .unwrap();
+fn test_package_logs() -> Result<()> {
+    let mut entries: Vec<DirEntry> =
+        fs::read_dir("./tests/package-logs")?.collect::<Result<Vec<_>, _>>()?;
     entries.sort_by_key(|e| e.file_name());
 
-    for entry in entries {
-        let file_contents = std::fs::read_to_string(entry.path()).unwrap();
-        let test: Test = serde_json::from_str(&file_contents).unwrap();
+    fs::create_dir_all("./tests/package-logs/output")?;
 
-        execute_test(test);
+    for entry in entries {
+        if entry.metadata()?.is_file() {
+            execute_test(&entry.path())?;
+        }
     }
-}
-
-fn execute_test(test: Test) {
-    let envelopes: Vec<Envelope<package::model::PackageRecord>> = test
-        .input
-        .into_iter()
-        .scan(None, |last, e_data| {
-            dbg!(e_data.contents.clone());
-            let key: signing::PrivateKey = e_data.key.parse().unwrap();
-            let mut record: package::model::PackageRecord = e_data.contents.try_into().unwrap();
-
-            record.prev = last.clone();
 
-            let envelope = Envelope::signed_contents(&key, record).unwrap();
+    Ok(())
+}
 
-            *last = Some(hash::HashAlgorithm::Sha256.digest(envelope.content_bytes()));
+fn validate_input(input: Vec<EnvelopeData>) -> Result<Validator> {
+    let mut validator = Validator::new();
+    for record in input.into_iter().scan(None, |last, e_data| {
```

Comment:
I think with a `scan` this big, it'll look better to chain it with a `fold` than make it the right-hand side of a loop.